crl-release-25.1: sstable: improve writer EstimatedSize

jbowens opened the pull request:
25.1 backport of https://github.com/cockroachdb/pebble/pull/4583 and https://github.com/cockroachdb/pebble/pull/4618.

RaduBerinde commented:
Oops, looks like there's a race:
```
WARNING: DATA RACE
                Write at 0x00c0003012b0 by goroutine 57:
                  github.com/cockroachdb/pebble/sstable.(*RawRowWriter).finishIndexBlock()
                      /home/runner/work/pebble/pebble/sstable/rowblk_writer.go:1382 +0x8fc
                  github.com/cockroachdb/pebble/sstable.(*RawRowWriter).addIndexEntry()
                      /home/runner/work/pebble/pebble/sstable/rowblk_writer.go:1243 +0x2fa
                  github.com/cockroachdb/pebble/sstable.(*writeQueue).performWrite()
                      /home/runner/work/pebble/pebble/sstable/write_queue.go:69 +0x3ed
                  github.com/cockroachdb/pebble/sstable.(*writeQueue).runWorker()
                      /home/runner/work/pebble/pebble/sstable/write_queue.go:100 +0xc4
                  github.com/cockroachdb/pebble/sstable.newWriteQueue.gowrap1()
                      /home/runner/work/pebble/pebble/sstable/write_queue.go:58 +0x33
                
                Previous read at 0x00c0003012b0 by goroutine 55:
                  github.com/cockroachdb/pebble/sstable.(*RawRowWriter).EstimatedSize()
                      /home/runner/work/pebble/pebble/sstable/rowblk_writer.go:1[72](https://github.com/cockroachdb/pebble/actions/runs/14666501487/job/41162486330?pr=4628#step:4:73)9 +0x127
                  github.com/cockroachdb/pebble/internal/compact.(*Runner).writeKeysToTable()
                      /home/runner/work/pebble/pebble/internal/compact/run.go:195 +0x99a
                  github.com/cockroachdb/pebble/internal/compact.(*Runner).WriteTable()
                      /home/runner/work/pebble/pebble/internal/compact/run.go:151 +0x26e
                  github.com/cockroachdb/pebble.(*DB).compactAndWrite()
                      /home/runner/work/pebble/pebble/compaction.go:3039 +0xec5
                  github.com/cockroachdb/pebble.(*DB).runCompaction()
                      /home/runner/work/pebble/pebble/compaction.go:2922 +0xe44
                  github.com/cockroachdb/pebble.(*DB).flush1()
                      /home/runner/work/pebble/pebble/compaction.go:14[82](https://github.com/cockroachdb/pebble/actions/runs/14666501487/job/41162486330?pr=4628#step:4:83) +0x88c
                  github.com/cockroachdb/pebble.(*DB).flush.func1()
                      /home/runner/work/pebble/pebble/compaction.go:1216 +0x124
```

jbowens commented:
Ah, right, I removed the possibility of parallelism here in #4418 but 25.1 predates that change. I switched this to just use an atomic. I don't believe we use the parallelism in Cockroach in production so it's a bit unfortunate to pay the atomic overhead :/

RaduBerinde reviewed (commented):
I don't think it matters, `Load()` is a regular read, and we only `Add` once per data block.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/4628#-:-OOibp7DDZ6vuDQj4qz_:b-nldi5w)* status: 0 of 10 files reviewed, all discussions resolved (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->
